File: src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

class MyNode(Node):
    def __init__(self):
        super().__init__("py_test")
        
        self.get_logger().info("Hello world")
        self.create_timer(1.0, self.timer_callback) # pass the reference to the function
        self.ctr = 1
        
    def timer_callback(self):
        self.get_logger().info("Hello" + " " + str(self.ctr))
        self.ctr += 1

def main(args = None):
    rclpy.init(args=args)
    node = MyNode()
    rclpy.spin(node) # this will allow the node to prcoess any callback
    # callback seems to be triggered only when node is spinning
    rclpy.shutdown()
    
    
if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"


class MyNode : public rclcpp::Node 
{
public:
    MyNode() : Node("cpp_test"){
        RCLCPP_INFO(this->get_logger(), "Hello World!");
        
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1), 
            std::bind(&MyNode::timer_callback, this)
        );
    }

private:
    rclcpp::TimerBase::SharedPtr timer_;
    int ctr = 1;

    
    void timer_callback(){
        RCLCPP_INFO(this->get_logger(), "Hello %d", this->ctr);
        (this->ctr)++;
    }
};
int main(int argc, char **argv){
    rclcpp::init(argc, argv);
    // in Cpp, we should create a shared pointer to the node class.
    // for reference, the shared pointer can be referred to by multiple entities.
    // we have multiple pointers that point to the same reference. 
    // maintains an initial reference counter.
    // if the shared pointer is copied, the counter is incremented.
    // the underlying resource is only deleted when the reference counter
    // hits zero.
    auto node = std::make_shared<MyNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}